src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -130,6 +130,37 @@ static void pcm_convert_s32_to_s24(struct comp_buffer *source,
 
 #endif /* CONFIG_FORMAT_S24LE && CONFIG_FORMAT_S32LE */
 
+#ifdef CONFIG_FORMAT_FLOAT && (CONFIG_FORMAT_S16LE || CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE)
+
+static inline int32_t _pcm_shift_left(int32_t d, int32_t a);
+static int32_t _pcm_convert_f_to_i(int32_t src);
+static int32_t _pcm_convert_i_to_f(int32_t src);
+
+static inline int32_t _pcm_shift_left(int32_t d, int32_t a)
+{
+	return (a >= 0) ? (d << a) : (d >> -a);
+}
+
+static int32_t _pcm_convert_f_to_i(int32_t src)
+{
+	int32_t xi, si, dst;
+	enum {mantisa_mask = (1<<23) - 1};
+
+	xi = ((src >> 23) & 0xFF) - 127; /* exponential */
+	si = ((mantisa_mask + 1) | (src & mantisa_mask)); /* mantisa + 1 */
+	si = (si ^ ( src >> 31 )) + (int)( (unsigned int)src >> 31); /* copy sign to mantisa */
+	dst = _pcm_shift_left(si, xi - 23); /* calculate power */
+	dst += _pcm_shift_left(si, xi - 22) & 1; /* add 0.5 to round correctly */
```

Comment:
...after we clarify what you meant to do with the sign extraction above, we can double-check this rounding for negative numbers.

Reply:
Also be careful with overflow when doing rounding (good that you included it), if need int64_t intermediate result and saturation is not expensive.